app/heron/factories/tube_rack.rb

```diff
@@ -9,28 +9,40 @@ class TubeRack
       include Concerns::HeronStudy
       include Concerns::CoordinatesSupport
       include Concerns::ForeignBarcodes
+      include Concerns::RecipientsCoordinates
+      include Concerns::Recipients
+      include Concerns::Contents
 
-      attr_accessor :sample_tubes, :tube_rack, :size
+      attr_accessor :sample_tubes, :tube_rack
 
-      validates_presence_of :tubes, :size, :purpose
-      validate :check_tubes
+      validates_presence_of :size, :purpose, :recipients
 
-      def tubes=(attributes)
-        attributes.each do |tube|
-          tubes[tube[:coordinate]] = ::Heron::Factories::Tube.new(tube.except(:coordinate).merge(study: heron_study))
-        end
+      def initialize(params)
+        @params = params
       end
 
-      def racked_tubes
-        @racked_tubes ||= []
+      def recipients_key
+        :tubes
       end
 
-      def purpose
-        @purpose ||= ::TubeRack::Purpose.where(target_type: 'TubeRack', size: size).first
+      def recipient_factory
+        ::Heron::Factories::Tube
+      end
+
+      def content_factory
+        ::Heron::Factories::Sample
+      end
+
+      def barcode
+        @params[:barcode]
+      end
+
+      def size
+        @params[:size]
```

Comment:
Does this method need to be changed to query the size from the Purpose record?

Reply:
Changed to get it from purpose